### PR TITLE
Feature/canonicalize column names

### PR DESCRIPTION
**Pull request by AlexanderMann**
# Motivation
https://github.com/datamill-co/target-postgres/issues/18

This is the first pr in a series. Proper handling of PSQL names involves non-latin characters, unicode, and general madness. As such, I've tried to split this work up into the following passes:

- detect invalid names (ASCII only)
- normalize names (leverage PSQL comments to store information about schema etc.)
- quotes support (?)
- unicode support (?)

## Suggested Musical Pairing
https://soundcloud.com/starfucker_usa/golden-light-1

### CODE REVIEW

**Comment by AlexanderMann** [comment]
@awm33 before I went ahead and did it...I figured I'd get your input. I'm thinking that all of the denesting of schema stuffs could be moved out either to `json_schema` or to `postgres_schema`. It'd greatly clean up the main `postgres` file.

File: target_postgres/postgres_schema.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+"""
+NAMEDATALEN _defaults_ to 64 in PostgreSQL. The maxmimum length for an identifier is
+  NAMEDATALEN - 1.
+
+  TODO: Figure out way to `SELECT` value from commands
+"""
+NAMEDATALEN = 63
+
+
+class SchemaError(Exception):
```
Comment (awm33) [inline target_postgres/postgres_schema.py]: Maybe `PostgresSchemaError` would be more clear in a stack trace?

File: target_postgres/postgres_schema.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                len(column_name),
+                NAMEDATALEN))
+
+    ## Subsequent characters in an identifier or key word can be letters, underscores, digits (0-9), or dollar signs ($)
+
+    lowered_name = column_name.lower()
+    return lowered_name[0] + re.sub(r'[^\w\d_$]', '_', lowered_name[1:])
+
+
+def add_singer_columns(schema, key_properties):
```
Comment (awm33) [inline target_postgres/postgres_schema.py]: This function doesn't have any postgres specific code.

**Review by awm33 — CHANGES_REQUESTED** [review]

**Comment by awm33** [comment]
@AlexanderMann I'm not sure if it's the right direction. I'm not a huge OO person, but creating a SQL target OO interface seems like the right direction and this PR is moving away from that. Clojure doesn't really have OO, but you could think of this as developing a protocol for SQL-like targets.

**Comment by AlexanderMann** [comment]
Makes sense. For what it's worth, I have been wanting to head in that
direction for a while, but have also wanted to not slow down velocity on
the task at hand for pure refactors etc.

I'm happy to do the legwork in this pr or close this and do that first.
Thoughts?

**Review by bender-peliqan — COMMENTED** [review]
## Code Review by Bender (automated)

Adds DELETERECORD support to target-postgres: `target_tools._line_handler` routes the new message type to `StreamTracker.handle_delete_record_message`, which buffers deletes per stream and flushes them **after** the record batch (so upsert+delete of the same PK in one run correctly ends up deleted). The actual DELETE is built as `DELETE FROM schema.table WHERE (pk_cols) IN ((%s,..),(%s,..),...)`, batched at `DELETE_BATCH_SIZE=1000`, with source-cased PK keys canonicalized to physical column names, deterministic skips (missing table / missing PK column) that rollback + return 0, and a `write_batch`-style transient-connection retry loop that raises on real failures. Overall: clean design, good separation, thoughtful test coverage.

### Looks good (verified)
- SQL is fully parameterised (`sql.Identifier`, `sql.Placeholder`); no interpolation of user data.
- Ordering guarantee (records flushed before deletes) is enforced in both `flush_stream` and `flush_streams`, and asserted in `test_records_flushed_before_deletes`.
- Watermark bookkeeping for deletes mirrors record semantics (`stream_add_watermarks`/`stream_flush_watermarks`), so STATE won't emit past an un-flushed delete.
- Canonicalisation of PK keys (`ID` -> `id`, `Account Id` -> `account_id`) closes a real silent no-op path — covered by `test_delete_key_is_canonicalized_to_physical_column` and `test_delete_key_with_invalid_chars_canonicalized`.
- Deterministic skips (missing table, missing PK column) rollback and return 0 rather than raising — correct best-effort behavior; both covered.
- Transient-connection retry loop raises on exhaustion instead of returning 0 (which would otherwise advance the delete bookmark past unlanded deletes) — asserted in `test_connection_loss_retries_then_raises`.
- `_flush_deletes` uses `hasattr(self.target, 'delete_records')` fallback — safe for any non-Postgres target reusing StreamTracker.
- Tests use the real `assert` keyword throughout, and each test has a `# WHY:` comment stating the invariant it protects.

### Concerns / questions

1. **PK-key uniformity assumption.** In `delete_records`:
   ```python
   raw_pk_columns = list(records[0].keys())
   ...
   params = [record[c] for record in batch for c in raw_pk_columns]
   ```
   The PK column set (and iteration order) is taken from `records[0]` and applied to every record in the batch. If a later record has a different key set or ordering (e.g. tap emits `{id, _sdc_division}` for one row and `{_sdc_division, id}` for the next), the second row either KeyErrors — raised as `PostgresError`, aborting the whole batch — or, worse, if the key set matches but the order is different, binds values to the wrong columns and silently deletes the wrong rows. In practice a well-behaved tap emits a stable PK shape per stream, but a defensive assert (`all(set(r.keys()) == set(raw_pk_columns) for r in batch)`) would turn a subtle data-integrity failure into a loud one. Worth adding, or at least a comment stating the invariant the caller must honour.

2. **Partial-PK-missing skips the entire batch.** In the missing-column branch, if the delete records carry keys `[id, division]` and only `id` exists on the table, the whole batch is silently skipped (log warning + return 0). That's the safe choice from a data-loss standpoint, but it means a mild schema drift on a *non-PK* helper column drops deletes on the floor. The log line is clear about *which* column is missing — good — but consider whether this should be an error rather than a skip when the missing column is not itself a PK. Judgement call; documenting the intended semantic in the docstring would be enough.

3. **DELETERECORD for an un-synced stream via `add_table_mapping_helper`.** `test_delete_for_unsynced_stream_is_routed` verifies the routing but stubs the DB entirely. `setup_table_mapping_cache(cur)` presumably repopulates the mapping from the live DB, so a table that exists but was never registered in this process should resolve — but I'd like to see this exercised in the live `test_postgres.py` suite (a delete-only sync of a pre-existing table) before rolling into a dedicated delete-sync run.

4. **`_flush_deletes` bumps the watermark even when the target doesn't implement `delete_records`.**
   ```python
   if hasattr(self.target, 'delete_records'):
       self.target.delete_records(stream, records)
   else:
       LOGGER.warning("Target does not implement delete_records; ...")
   self.stream_flush_watermarks[stream] = self.stream_add_watermarks.get(...)
   ```
   The watermark advance is unconditional. For target-postgres this branch is dead, but any other target reusing `StreamTracker` (target-bigquery, per the companion PR) will silently drop deletes AND emit STATE past them. Consider guarding the watermark bump on the `if` branch, or moving it inside — otherwise the "safe fallback" quietly corrupts the delete bookmark.

### Test coverage

Coverage is thorough at the unit level — routing, ordering, batching, canonicalisation, composite PK, empty input, missing table, missing PK column, execute failure (raises), transient-connection retry (exhaustion + recovery). Two follow-ups worth considering:

- A test with a batch whose second record has a different key set (asserts either a clean raise or the invariant — see concern #1).
- Live-DB coverage in `test_postgres.py` for the un-synced-stream path (concern #3).

### Minor
- `postgres.py` uses `##` (double comment marker) for the `DELETE_BATCH_SIZE` class-level comment — the rest of the file uses `#`. Cosmetic.
- `deleted_total += cur.rowcount if cur.rowcount and cur.rowcount > 0 else 0` — the `cur.rowcount and cur.rowcount > 0` is redundant; `max(cur.rowcount or 0, 0)` reads cleaner. Nit.
- Docstring on `delete_records` is unusually good (retry semantics, skip semantics, canonicalisation rationale all called out). Keep this bar.

### Recommendation
Comment / approve-with-suggestions. The core delete pipeline is well-designed, the SQL layer is parameterised, and the tests genuinely protect the invariants that matter (ordering, canonicalisation, no-silent-swallowing-of-real-failures). Nothing here is blocking, but I'd address concern #4 (unconditional watermark advance in the no-delete-support branch) before this ships alongside the target-bigquery companion PR — that's the branch most likely to bite in practice.


---
_Warning: no reviewers are currently assigned to this PR._

@AlexanderMann FYI — automated review posted.

**Pull request closed without merging**